### PR TITLE
Feat/trade field

### PR DESCRIPTION
**Pull request by TarasBlatnoi**

### CODE REVIEW

File: src/auth/guard/turn.guard.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+export class TurnGuard implements CanActivate {
+  constructor(private gameService: GameService) {}
+
+  async canActivate(context: any): Promise<boolean | any> {
+    const client = context.switchToWs().getClient();
+    const userId = client.jwtPayload.sub;
+    const { gameId } = parse(client.handshake.headers.cookie);
+    try {
+      const game = await this.gameService.getCurrentGame(gameId);
+      if (game.turnOfUserId !== userId) throw new WsException('Wrong turn');
```
Comment (Moriartymath) [inline src/auth/guard/turn.guard.ts:16]: Use optional chaining in case of non-existing game

File: src/game/game.service.ts [hunk 2]
```diff
@@ -45,8 +54,12 @@ export class GameService {
     );
     if (filteredPlayers.length || game.status === 'ACTIVE')
```
Comment (Moriartymath) [inline src/game/game.service.ts:55]: It`s better to use 'some' method of Array, if you are only checking for user presence

File: src/game/game.service.ts [hunk 3]
```diff
@@ -94,26 +107,61 @@ export class GameService {
...
+      this.logger.log(`Timer with id:${id} was set for ${time / 1000} seconds`);
+    });
+  }
 
-    this.logger.log(
-      `User ${game.turnOfUserId} has 30 seconds to roll the dice in game ${game.id}.`
+  calculateEndOfTurn(timeOfTurn: number) {
+    let turnEnds = Date.now();
+    turnEnds += timeOfTurn;
+    return turnEnds.toString();
```
Comment (Moriartymath) [inline src/game/game.service.ts:142]: why are you transform number to string?

File: src/player/dto/create-player.dto.ts [hunk 4]
```diff
@@ -1,8 +1,10 @@
 import { IsUUID } from 'class-validator';
+import { FieldsType } from 'src/utils/fields';
 
 export class CreatePlayerDto {
   @IsUUID()
   gameId: string;
   @IsUUID()
   userId: string;
+  color: string;
```
Comment (Moriartymath) [inline src/player/dto/create-player.dto.ts:9]: add validator for color filed - IsString()

File: src/game/game.service.ts [hunk 5]
```diff
@@ -125,26 +173,184 @@ export class GameService {
...
+      currentPlayer.currentFieldIndex
+    );
+    this.deletePlayerId(playerField.players, currentPlayer.id);
+    const playerNextField = this.findPlayerFieldByIndex(fields, nextIndex);
+    playerNextField.players.push(updatedPlayer.id);
+    return { updatedGame, fields, nextIndex, playerNextField };
+  }
+
+  async putUpForAuction(game: Partial<GamePayload>) {
+    const player = await this.playerService.findByUserAndGameId(
```
Comment (Moriartymath) [inline src/game/game.service.ts:267]: you have to check for null to be returned

File: src/game/game.service.ts [hunk 6]
```diff
@@ -125,26 +173,184 @@ export class GameService {
...
+    this.setAuction(gameId, auction);
+    return auction;
+  }
+
+  async raisePrice(gameId: string, userId: string, raiseBy: number) {
+    const player = await this.playerService.findByUserAndGameId(userId, gameId);
+    const auction = this.getAuction(gameId);
+    if (raiseBy < 100) throw new WsException('Raise is not big enough');
+    if (!auction) throw new WsException('Auction wasn’t started');
+    if (player.money < auction.bid) throw new WsException('Not enough money');
```
Comment (Moriartymath) [inline src/game/game.service.ts:302]: you check for user to be null

File: src/game/game.service.ts (hunk 3 above)
Comment (TarasBlatnoi) [reply src/game/game.service.ts:142]: In db it is saved as string because we dont want to play with BigInt

**Pull request merged**